File: src/code_indexing_mcp/installer/shell_path.py
"""The ``code-indexing-mcp`` launcher and the PATH entry that makes it reachable.

The server executable lives inside the installation's virtual environment, and
that absolute path is what every harness config names. A shell has no way to
find it, so the CLI -- ``configure``, ``index``, ``status``, ``daemon`` -- is
unreachable without help. This module puts a launcher in a bin directory and,
when that directory is not already on PATH, offers to add it to the user's
shell profile.

The launcher is a symlink: the venv console script carries an absolute shebang,
so it runs correctly through a link with no wrapper of any kind. Windows is the
exception -- creating a symlink there needs a privilege the installer cannot
assume -- so it gets a ``.cmd`` shim instead.

No Textual import belongs here; the TUI is one caller of this, not its owner.
"""

from __future__ import annotations

import contextlib
import os
import shutil
import string
import sys
from collections.abc import Mapping, Sequence
from dataclasses import dataclass
from pathlib import Path

from .accelerator import server_executable
from .config_files import InstallerError, write_changed_configuration
from .links import backup_path, is_under, link_destination, replace_link

LAUNCHER_NAME = "code-indexing-mcp"

# The block written into a shell profile. Both markers are matched literally on
# rewrite and removal, which is what keeps a second install from appending a
# second copy and lets the uninstaller take back exactly what it added.
BLOCK_START = "# >>> code-indexing-mcp >>>"
BLOCK_END = "# <<< code-indexing-mcp <<<"


@dataclass(frozen=True)
class LauncherResult:
    """What happened to the launcher, and where it ended up."""

    path: Path
    # "created" | "current" | "replaced" | "skipped" | "failed"
    status: str
    detail: str = ""

    @property
    def ok(self) -> bool:
        return self.status in {"created", "current", "replaced"}


@dataclass(frozen=True)
class PathState:
    """Everything the wizard needs to describe the PATH situation up front."""

    bin_directory: Path
    launcher: Path
    on_path: bool
    # A `code-indexing-mcp` that an earlier PATH entry would find first.
    shadowed_by: Path | None
    # Profiles that would be edited if the user asks for it.
    profiles: tuple[Path, ...]
    # True when every profile already carries the block or the directory.
    profiles_current: bool


def default_bin_directory(
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
) -> Path:
    """Where the launcher goes: the XDG user binary directory by convention."""

    home = home or Path.home()
    environment = os.environ if environment is None else environment
    for variable in ("CODE_INDEXING_MCP_BIN_DIR", "XDG_BIN_HOME"):
        configured = environment.get(variable)
        if configured:
            return Path(configured).expanduser()
    return home / ".local" / "bin"


def launcher_path(bin_directory: Path, *, platform_name: str | None = None) -> Path:
    platform_name = platform_name or sys.platform
    if platform_name.startswith("win"):
        return bin_directory / f"{LAUNCHER_NAME}.cmd"
    return bin_directory / LAUNCHER_NAME


def _path_entries(environment: Mapping[str, str]) -> list[Path]:
    raw = environment.get("PATH", "")
    entries: list[Path] = []
    for part in raw.split(os.pathsep):
        if not part:
            continue
        try:
            entries.append(Path(part).expanduser())
        except (OSError, ValueError):
            # A malformed PATH entry is the user's problem, not a reason to
            # abandon the check; the remaining entries still answer the question.
            continue
    return entries


def _same_directory(left: Path, right: Path) -> bool:
    """Compare two directories without requiring either to exist."""

    try:
        return left.resolve() == right.resolve()
    except (OSError, ValueError):
        # An entry the filesystem refuses to stat at all -- a NUL byte, a path
        # past the length limit -- is still comparable as text.
        return left == right


def is_on_path(
    bin_directory: Path,
    *,
    environment: Mapping[str, str] | None = None,
) -> bool:
    environment = os.environ if environment is None else environment
    return any(_same_directory(entry, bin_directory) for entry in _path_entries(environment))


def shadowing_executable(
    bin_directory: Path,
    *,
    environment: Mapping[str, str] | None = None,
    platform_name: str | None = None,
) -> Path | None:
    """A ``code-indexing-mcp`` an earlier PATH entry would find before ours.

    Reported rather than removed. Something else owns that file, and silently
    winning the name would be a worse surprise than saying which one runs.
    """

    environment = os.environ if environment is None else environment
    ours = launcher_path(bin_directory, platform_name=platform_name)
    for entry in _path_entries(environment):
        if _same_directory(entry, bin_directory):
            return None
        candidate = shutil.which(LAUNCHER_NAME, path=str(entry))
        if candidate is not None:
            found = Path(candidate)
            if not _same_directory(found.parent, ours.parent):
                return found
    return None


def install_launcher(
    install_directory: Path,
    bin_directory: Path,
    *,
    platform_name: str | None = None,
) -> LauncherResult:
    """Put a ``code-indexing-mcp`` launcher in ``bin_directory``.

    Never raises: a launcher is a convenience, and a failure to create one must
    not undo harness configuration that already succeeded.
    """

    platform_name = platform_name or sys.platform
    target = launcher_path(bin_directory, platform_name=platform_name)
    executable = server_executable(install_directory, platform_name=platform_name)
    if not executable.is_file():
        return LauncherResult(
            target,
            "failed",
            f"no server executable at {executable}",
        )
    try:
        if platform_name.startswith("win"):
            return _install_shim(executable, target)
        return _install_symlink(executable, target)
    except (OSError, UnicodeDecodeError) as exc:
        return LauncherResult(target, "failed", str(exc))


def _is_our_launcher(target: Path, executable: Path) -> bool:
    """True when ``target`` is a link this installer left pointing at a venv of ours."""

    if not target.is_symlink():
        return False
    destination = link_destination(target)
    return destination.name == executable.name and ".venv" in destination.parts


def _install_symlink(executable: Path, target: Path) -> LauncherResult:
    stale = _is_our_launcher(target, executable)
    # Resolve the backup name before the move, not after: once replace_link has
    # renamed the old entry, backup_path returns the *next* free name instead of
    # the one the user's file is now under.
    backup = backup_path(target) if (target.is_symlink() or target.exists()) and not stale else None
    created = replace_link(executable, target, is_directory=False, stale=stale)
    if not created:
        return LauncherResult(target, "current", f"already points at {executable}")
    if backup is not None:
        return LauncherResult(target, "replaced", f"the previous entry was kept as {backup.name}")
    return LauncherResult(target, "created", f"points at {executable}")


_SHIM_TEMPLATE = '@echo off\r\n"{executable}" %*\r\n'


def _install_shim(executable: Path, target: Path) -> LauncherResult:
    """Windows gets a batch shim: symlinks there need a privilege we cannot assume."""

    content = _SHIM_TEMPLATE.format(executable=executable)
    # Read as bytes: the shim's CRLF line endings are written verbatim, and
    # universal-newline decoding would turn every re-install into a rewrite.
    existing = target.read_bytes().decode("utf-8") if target.is_file() else None
    if existing == content:
        return LauncherResult(target, "current", f"already runs {executable}")
    target.parent.mkdir(parents=True, exist_ok=True)
    changed = write_changed_configuration(target, existing, content)
    if not changed:  # pragma: no cover - equality was already handled above
        return LauncherResult(target, "current", f"already runs {executable}")
    status = "replaced" if existing is not None else "created"
    return LauncherResult(target, status, f"runs {executable}")


def _removable_launcher(
    target: Path,
    install_directory: Path | None,
    platform_name: str,
) -> bool:
    """True when ``target`` is evidently a launcher this installation created.

    ``.venv`` in the destination is necessary but nowhere near sufficient: a
    user's own symlink to some unrelated project's virtual environment matches
    it too. When the caller knows which checkout is being uninstalled -- and the
    uninstaller always does -- the destination has to point inside that one.
    """

    if platform_name.startswith("win"):
        if not target.is_file():
            return False
        try:
            content = target.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            return False
        if ".venv" not in content or LAUNCHER_NAME not in content:
            return False
        if install_directory is None:
            return True
        # The shim names the executable by absolute path on its one command line.
        return any(
            is_under(Path(candidate), install_directory)
            for candidate in _shim_command_candidates(content)
        )
    if not target.is_symlink():
        return False
    destination = link_destination(target)
    if ".venv" not in destination.parts:
        return False
    if install_directory is None:
        return True
    return is_under(destination, install_directory)


def _shim_command_candidates(content: str) -> list[str]:
    """The quoted executable paths a Windows shim names, if it names any."""

    candidates: list[str] = []
    for line in content.splitlines():
        opening = line.find('"')
        closing = line.find('"', opening + 1) if opening != -1 else -1
        if closing != -1:
            candidates.append(line[opening + 1 : closing])
    return candidates


def remove_launcher(
    bin_directory: Path,
    install_directory: Path | None = None,
    *,
    platform_name: str | None = None,
) -> Path | None:
    """Remove a launcher this installer created. Returns the path if one went away.

    A file at that name that is *not* one of ours is left strictly alone. Pass
    ``install_directory`` -- the checkout being removed -- to require that the
    launcher actually points into it; without it the check falls back to the
    looser "points into some virtual environment" test.
    """

    platform_name = platform_name or sys.platform
    target = launcher_path(bin_directory, platform_name=platform_name)
    if not _removable_launcher(target, install_directory, platform_name):
        return None
    target.unlink()
    return target


def _fish_profile(home: Path, environment: Mapping[str, str]) -> Path:
    configured = environment.get("XDG_CONFIG_HOME")
    base = Path(configured).expanduser() if configured else home / ".config"
    return base / "fish" / "config.fish"


def shell_profiles(
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
    platform_name: str | None = None,
) -> tuple[Path, ...]:
    """The profile files to edit, in the order they would be written.

    ``$SHELL`` names the login shell, which decides the primary file; any other
    profile that already exists is included too, so a user who moves between
    bash and zsh does not lose the entry by switching.
    """

    home = home or Path.home()
    environment = os.environ if environment is None else environment
    platform_name = platform_name or sys.platform
    if platform_name.startswith("win"):
        # A profile edit on Windows means the registry or `setx`; the installer
        # reports the manual step instead of reaching for either.
        return ()

    zdotdir = environment.get("ZDOTDIR")
    zshrc = (Path(zdotdir).expanduser() if zdotdir else home) / ".zshrc"
    bashrc = home / ".bashrc"
    # macOS terminals start bash as a login shell, which reads .bash_profile and
    # never .bashrc unless that file sources it.
    bash_profile = home / ".bash_profile"
    fish = _fish_profile(home, environment)
    profile = home / ".profile"

    shell = Path(environment.get("SHELL", "")).name
    primary: list[Path] = []
    if shell == "zsh":
        primary = [zshrc]
    elif shell == "bash":
        primary = [bashrc, bash_profile] if platform_name == "darwin" else [bashrc]
    elif shell == "fish":
        primary = [fish]
    elif shell:
        primary = [profile]

    selected: list[Path] = list(primary)
    for candidate in (zshrc, bashrc, bash_profile, fish, profile):
        if candidate not in selected and candidate.is_file():
            selected.append(candidate)
    if not selected:
        selected = [profile]
    return tuple(selected)


# What a double-quoted word still interprets, per shell. POSIX shells run
# command substitution inside double quotes; fish does not, and escaping a
# backtick there would leave a literal backslash in the path.
_POSIX_SPECIALS = ("\\", '"', "`", "$")
_FISH_SPECIALS = ("\\", '"', "$")


def _escaped(text: str, specials: tuple[str, ...]) -> str:
    """Escape ``text`` for use inside a double-quoted shell word."""

    for character in specials:  # the backslash comes first, or it doubles the rest
        text = text.replace(character, f"\\{character}")
    return text


def _quoted_location(bin_directory: Path, home: Path, specials: tuple[str, ...]) -> str:
    """The directory as the inside of a double-quoted word, ``$HOME`` still live.

    A bin directory is not always a tame path: ``--bin-dir`` takes whatever the
    user types, and a space, a quote, or a ``$(`` reaching a shell profile
    unescaped would either break every future shell or run something.
    """

    try:
        relative = bin_directory.relative_to(home).as_posix()
    except ValueError:
        return _escaped(str(bin_directory), specials)
    return "$HOME/" + _escaped(relative, specials)


def _block(bin_directory: Path, profile: Path, home: Path) -> str:
    """The marked block for one profile, in that profile's own syntax."""

    if profile.name == "config.fish":
        line = f'fish_add_path "{_quoted_location(bin_directory, home, _FISH_SPECIALS)}"'
    else:
        line = f'export PATH="{_quoted_location(bin_directory, home, _POSIX_SPECIALS)}:$PATH"'
    return f"{BLOCK_START}\n{line}\n{BLOCK_END}\n"


def _home_relative(path: Path, home: Path) -> str:
    """Render a path under home as ``$HOME/...`` so the line survives a moved home."""

    try:
        relative = path.relative_to(home)
    except ValueError:
        return str(path)
    return f"$HOME/{relative.as_posix()}"


# Characters that continue a path token. A needle followed by one of these is a
# prefix of some other directory, not the directory itself.
_PATH_CHARACTERS = frozenset(string.ascii_letters + string.digits + "-_./~$")


def _mentions_whole_path(line: str, needle: str) -> bool:
    """True when ``needle`` appears in ``line`` as a complete path token.

    A plain substring test reports ``~/bin`` as present in a line that only
    mentions ``~/bin2``, which would leave the user with no PATH entry and an
    install that says their profile already had one.
    """

    start = line.find(needle)
    while start != -1:
        end = start + len(needle)
        before = start == 0 or line[start - 1] not in _PATH_CHARACTERS
        after = end >= len(line) or line[end] not in _PATH_CHARACTERS
        if before and after:
            return True
        start = line.find(needle, start + 1)
    return False


def profile_mentions_directory(text: str, bin_directory: Path, home: Path) -> bool:
    """True when a profile already puts ``bin_directory`` on PATH somehow.

    Deliberately loose about *how* the user wrote it -- our own block, an
    ``export PATH`` line, a ``$HOME``- or ``~``-relative spelling all count --
    but strict about the path itself, so a neighbouring directory whose name
    starts the same way does not silently suppress the entry.
    """

    if BLOCK_START in text:
        return True
    needles = {str(bin_directory), _home_relative(bin_directory, home)}
    with contextlib.suppress(ValueError):  # not under home; the two above suffice
        needles.add(f"~/{bin_directory.relative_to(home).as_posix()}")
    return any(
        _mentions_whole_path(line, needle) for line in text.splitlines() for needle in needles
    )


def update_profile(
    profile: Path,
    bin_directory: Path,
    *,
    home: Path | None = None,
) -> bool:
    """Append the marked block to one profile. Returns True when it was written."""

    home = home or Path.home()
    try:
        original: str | None = profile.read_text(encoding="utf-8")
    except FileNotFoundError:
        original = None
    except UnicodeDecodeError as exc:
        # A profile in some other encoding cannot be safely appended to:
        # rewriting it would decide an encoding on the user's behalf. This is a
        # failure, not a no-op -- returning False here would be indistinguishable
        # from "already configured", and the run would claim a PATH entry it
        # never wrote.
        raise InstallerError(f"{profile} is not valid UTF-8, so it was left alone") from exc
    # Any other OSError -- a permission wall, a directory where a file belongs --
    # is a real failure and travels up to update_profiles, which reports it.
    if original is not None and profile_mentions_directory(original, bin_directory, home):
        return False
    prefix = "" if not original else ("" if original.endswith("\n") else "\n")
    updated = (original or "") + prefix + _block(bin_directory, profile, home)
    return write_changed_configuration(profile, original, updated)


def update_profiles(
    bin_directory: Path,
    profiles: Sequence[Path],
    *,
    home: Path | None = None,
) -> tuple[list[Path], list[tuple[Path, str]]]:
    """Append the block to each profile, keeping one file's failure isolated."""

    written: list[Path] = []
    failures: list[tuple[Path, str]] = []
    for profile in profiles:
        try:
            if update_profile(profile, bin_directory, home=home):
                written.append(profile)
        except (OSError, InstallerError) as exc:
            failures.append((profile, str(exc)))
    return written, failures


def remove_path_block(profile: Path) -> bool:
    """Remove every marked block from one profile. True when one was there."""

    try:
        original = profile.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return False
    updated = original
    while True:
        start = updated.find(BLOCK_START)
        if start == -1:
            break
        end = updated.find(BLOCK_END, start)
        if end == -1:
            # A start marker with no end marker means the user edited the block
            # by hand. Removing to end-of-file would take their edit with it.
            break
        end += len(BLOCK_END)
        if updated[end : end + 1] == "\n":
            end += 1
        updated = updated[:start] + updated[end:]
    return write_changed_configuration(profile, original, updated)


def inspect(
    bin_directory: Path,
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
    platform_name: str | None = None,
) -> PathState:
    """Describe the PATH situation for a bin directory, for display before writing."""

    home = home or Path.home()
    environment = os.environ if environment is None else environment
    on_path = is_on_path(bin_directory, environment=environment)
    profiles = shell_profiles(home=home, environment=environment, platform_name=platform_name)
    current = True
    for profile in profiles:
        try:
            text = profile.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            current = False
            continue
        if not profile_mentions_directory(text, bin_directory, home):
            current = False
    return PathState(
        bin_directory=bin_directory,
        launcher=launcher_path(bin_directory, platform_name=platform_name),
        on_path=on_path,
        shadowed_by=shadowing_executable(
            bin_directory,
            environment=environment,
            platform_name=platform_name,
        ),
        profiles=profiles,
        profiles_current=current,
    )


def activation_hint(
    profiles: Sequence[Path],
    *,
    environment: Mapping[str, str] | None = None,
) -> str:
    """The command that makes a freshly written PATH entry live in this shell."""

    environment = os.environ if environment is None else environment
    if any(profile.name == "config.fish" for profile in profiles):
        return "exec fish"
    shell = environment.get("SHELL")
    return f"exec {shell} -l" if shell else "exec $SHELL -l"

File: ts/packages/installer/src/shell-path.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { serverExecutable } from "./accelerator.ts";
import { expandUser, InstallerError, writeChangedConfiguration } from "./config-files.ts";
import { backupPath, isUnder, linkDestination, replaceLink } from "./links.ts";

export const LAUNCHER_NAME = "code-indexing-mcp";
export const BLOCK_START = "# >>> code-indexing-mcp >>>";
export const BLOCK_END = "# <<< code-indexing-mcp <<<";

export interface LauncherResult {
  readonly path: string;
  readonly status: "created" | "current" | "replaced" | "skipped" | "failed";
  readonly detail: string;
}

export function launcherOk(result: LauncherResult): boolean {
  return result.status === "created" || result.status === "current" || result.status === "replaced";
}

export interface PathState {
  readonly binDirectory: string;
  readonly launcher: string;
  readonly onPath: boolean;
  readonly shadowedBy: string | null;
  readonly profiles: readonly string[];
  readonly profilesCurrent: boolean;
}

export function defaultBinDirectory(
  options: { home?: string; environment?: NodeJS.ProcessEnv } = {},
): string {
  const home = options.home ?? os.homedir();
  const environment = options.environment ?? process.env;
  for (const variable of ["CODE_INDEXING_MCP_BIN_DIR", "XDG_BIN_HOME"]) {
    const configured = environment[variable];
    if (configured !== undefined && configured !== "") return expandUser(configured);
  }
  return path.join(home, ".local", "bin");
}

export function launcherPath(
  binDirectory: string,
  platformName: string = process.platform,
): string {
  if (platformName.startsWith("win")) return path.join(binDirectory, `${LAUNCHER_NAME}.cmd`);
  return path.join(binDirectory, LAUNCHER_NAME);
}

function pathEntries(environment: NodeJS.ProcessEnv): string[] {
  const raw = environment.PATH ?? environment.Path ?? "";
  const entries: string[] = [];
  for (const part of raw.split(path.delimiter)) {
    if (part === "") continue;
    try {
      entries.push(expandUser(part));
    } catch {}
  }
  return entries;
}

function sameDirectory(left: string, right: string): boolean {
  try {
    return path.resolve(left) === path.resolve(right);
  } catch {
    return left === right;
  }
}

export function isOnPath(
  binDirectory: string,
  options: { environment?: NodeJS.ProcessEnv } = {},
): boolean {
  const environment = options.environment ?? process.env;
  return pathEntries(environment).some((entry) => sameDirectory(entry, binDirectory));
}

function whichIn(name: string, directory: string): string | null {
  const candidates =
    process.platform === "win32" ? [name, `${name}.cmd`, `${name}.exe`, `${name}.bat`] : [name];
  for (const candidate of candidates) {
    const filePath = path.join(directory, candidate);
    try {
      if (fs.statSync(filePath).isFile()) return filePath;
    } catch {}
  }
  return null;
}

export function shadowingExecutable(
  binDirectory: string,
  options: { environment?: NodeJS.ProcessEnv; platformName?: string } = {},
): string | null {
  const environment = options.environment ?? process.env;
  const ours = launcherPath(binDirectory, options.platformName);
  for (const entry of pathEntries(environment)) {
    if (sameDirectory(entry, binDirectory)) return null;
    const found = whichIn(LAUNCHER_NAME, entry);
    if (found !== null && !sameDirectory(path.dirname(found), path.dirname(ours))) return found;
  }
  return null;
}

function isOurLauncher(target: string, executable: string): boolean {
  try {
    if (!fs.lstatSync(target).isSymbolicLink()) return false;
  } catch {
    return false;
  }
  const destination = linkDestination(target);
  const parts = destination.split(path.sep);
  return (
    path.basename(destination) === path.basename(executable) &&
    (parts.includes(".venv") || parts.includes("bin"))
  );
}

function installSymlink(executable: string, target: string): LauncherResult {
  const stale = isOurLauncher(target, executable);
  let backup: string | null = null;
  try {
    if ((fs.lstatSync(target).isSymbolicLink() || fs.existsSync(target)) && !stale) {
      backup = backupPath(target);
    }
  } catch {
    // Target does not exist.
  }
  const created = replaceLink(executable, target, { isDirectory: false, stale });
  if (!created) {
    return { path: target, status: "current", detail: `already points at ${executable}` };
  }
  if (backup !== null) {
    return {
      path: target,
      status: "replaced",
      detail: `the previous entry was kept as ${path.basename(backup)}`,
    };
  }
  return { path: target, status: "created", detail: `points at ${executable}` };
}

const SHIM_TEMPLATE = '@echo off\r\n"{executable}" %*\r\n';

function installShim(executable: string, target: string): LauncherResult {
  const content = SHIM_TEMPLATE.replace("{executable}", executable);
  let existing: string | null = null;
  try {
    if (fs.statSync(target).isFile()) existing = fs.readFileSync(target).toString("utf8");
  } catch {
    existing = null;
  }
  if (existing === content) {
    return { path: target, status: "current", detail: `already runs ${executable}` };
  }
  fs.mkdirSync(path.dirname(target), { recursive: true });
  const changed = writeChangedConfiguration(target, existing, content);
  if (!changed) {
    return { path: target, status: "current", detail: `already runs ${executable}` };
  }
  return {
    path: target,
    status: existing !== null ? "replaced" : "created",
    detail: `runs ${executable}`,
  };
}

export function installLauncher(
  installDirectory: string,
  binDirectory: string,
  options: { platformName?: string } = {},
): LauncherResult {
  const platformName = options.platformName ?? process.platform;
  const target = launcherPath(binDirectory, platformName);
  const executable = serverExecutable(installDirectory, platformName);
  try {
    if (!fs.statSync(executable).isFile()) {
      return { path: target, status: "failed", detail: `no server executable at ${executable}` };
    }
  } catch {
    return { path: target, status: "failed", detail: `no server executable at ${executable}` };
  }
  try {
    return platformName.startsWith("win")
      ? installShim(executable, target)
      : installSymlink(executable, target);
  } catch (error) {
    return {
      path: target,
      status: "failed",
      detail: error instanceof Error ? error.message : String(error),
    };
  }
}

function shimCommandCandidates(content: string): string[] {
  const candidates: string[] = [];
  for (const line of content.split(/\r?\n/)) {
    const opening = line.indexOf('"');
    const closing = opening === -1 ? -1 : line.indexOf('"', opening + 1);
    if (closing !== -1) candidates.push(line.slice(opening + 1, closing));
  }
  return candidates;
}

function removableLauncher(
  target: string,
  installDirectory: string | null,
  platformName: string,
): boolean {
  if (platformName.startsWith("win")) {
    try {
      if (!fs.statSync(target).isFile()) return false;
      const content = fs.readFileSync(target, "utf8");
      if (!content.includes("bin") || !content.includes(LAUNCHER_NAME)) return false;
      if (installDirectory === null) return true;
      return shimCommandCandidates(content).some((candidate) =>
        isUnder(candidate, installDirectory),
      );
    } catch {
      return false;
    }
  }
  try {
    if (!fs.lstatSync(target).isSymbolicLink()) return false;
  } catch {
    return false;
  }
  const destination = linkDestination(target);
  const parts = destination.split(path.sep);
  if (!parts.includes(".venv") && !parts.includes("bin")) return false;
  if (installDirectory === null) return true;
  return isUnder(destination, installDirectory);
}

export function removeLauncher(
  binDirectory: string,
  installDirectory: string | null = null,
  options: { platformName?: string } = {},
): string | null {
  const platformName = options.platformName ?? process.platform;
  const target = launcherPath(binDirectory, platformName);
  if (!removableLauncher(target, installDirectory, platformName)) return null;
  fs.unlinkSync(target);
  return target;
}

function fishProfile(home: string, environment: NodeJS.ProcessEnv): string {
  const configured = environment.XDG_CONFIG_HOME;
  const base =
    configured !== undefined && configured !== ""
      ? expandUser(configured)
      : path.join(home, ".config");
  return path.join(base, "fish", "config.fish");
}

export function shellProfiles(
  options: { home?: string; environment?: NodeJS.ProcessEnv; platformName?: string } = {},
): string[] {
  const home = options.home ?? os.homedir();
  const environment = options.environment ?? process.env;
  const platformName = options.platformName ?? process.platform;
  if (platformName.startsWith("win")) return [];

  const zdotdir = environment.ZDOTDIR;
  const zshrc = path.join(
    zdotdir !== undefined && zdotdir !== "" ? expandUser(zdotdir) : home,
    ".zshrc",
  );
  const bashrc = path.join(home, ".bashrc");
  const bashProfile = path.join(home, ".bash_profile");
  const fish = fishProfile(home, environment);
  const profile = path.join(home, ".profile");

  const shell = path.basename(environment.SHELL ?? "");
  let primary: string[] = [];
  if (shell === "zsh") primary = [zshrc];
  else if (shell === "bash") primary = platformName === "darwin" ? [bashrc, bashProfile] : [bashrc];
  else if (shell === "fish") primary = [fish];
  else if (shell !== "") primary = [profile];

  const selected = [...primary];
  for (const candidate of [zshrc, bashrc, bashProfile, fish, profile]) {
    if (
      !selected.includes(candidate) &&
      fs.existsSync(candidate) &&
      fs.statSync(candidate).isFile()
    ) {
      selected.push(candidate);
    }
  }
  if (selected.length === 0) selected.push(profile);
  return selected;
}

const POSIX_SPECIALS = ["\\", '"', "`", "$"] as const;
const FISH_SPECIALS = ["\\", '"', "$"] as const;

function escaped(text: string, specials: readonly string[]): string {
  let value = text;
  for (const character of specials) {
    value = value.split(character).join(`\\${character}`);
  }
  return value;
}

function quotedLocation(binDirectory: string, home: string, specials: readonly string[]): string {
  const relative = path.relative(home, binDirectory);
  if (relative.startsWith("..") || path.isAbsolute(relative)) {
    return escaped(binDirectory, specials);
  }
  return `$HOME/${escaped(relative.split(path.sep).join("/"), specials)}`;
}

function block(binDirectory: string, profile: string, home: string): string {
  const line =
    path.basename(profile) === "config.fish"
      ? `fish_add_path "${quotedLocation(binDirectory, home, FISH_SPECIALS)}"`
      : `export PATH="${quotedLocation(binDirectory, home, POSIX_SPECIALS)}:$PATH"`;
  return `${BLOCK_START}\n${line}\n${BLOCK_END}\n`;
}

function homeRelative(filePath: string, home: string): string {
  const relative = path.relative(home, filePath);
  if (relative.startsWith("..") || path.isAbsolute(relative)) return filePath;
  return `$HOME/${relative.split(path.sep).join("/")}`;
}

const PATH_CHARACTERS = new Set(
  "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-_./~$".split(""),
);

function mentionsWholePath(line: string, needle: string): boolean {
  let start = line.indexOf(needle);
  while (start !== -1) {
    const end = start + needle.length;
    const before = start === 0 || !PATH_CHARACTERS.has(line[start - 1] ?? "");
    const after = end >= line.length || !PATH_CHARACTERS.has(line[end] ?? "");
    if (before && after) return true;
    start = line.indexOf(needle, start + 1);
  }
  return false;
}

export function profileMentionsDirectory(
  text: string,
  binDirectory: string,
  home: string,
): boolean {
  if (text.includes(BLOCK_START)) return true;
  const needles = new Set([binDirectory, homeRelative(binDirectory, home)]);
  const relative = path.relative(home, binDirectory);
  if (!relative.startsWith("..") && !path.isAbsolute(relative)) {
    needles.add(`~/${relative.split(path.sep).join("/")}`);
  }
  return text
    .split(/\r?\n/)
    .some((line) => [...needles].some((needle) => mentionsWholePath(line, needle)));
}

export function updateProfile(
  profile: string,
  binDirectory: string,
  options: { home?: string } = {},
): boolean {
  const home = options.home ?? os.homedir();
  let original: string | null;
  try {
    original = fs.readFileSync(profile, "utf8");
  } catch (error) {
    if (error instanceof Error && "code" in error && error.code === "ENOENT") original = null;
    else if (error instanceof Error && error.message.includes("encoding")) {
      throw new InstallerError(`${profile} is not valid UTF-8, so it was left alone`);
    } else {
      throw error;
    }
  }
  if (original !== null && profileMentionsDirectory(original, binDirectory, home)) return false;
  const prefix = original === null || original === "" || original.endsWith("\n") ? "" : "\n";
  const updated = `${original ?? ""}${prefix}${block(binDirectory, profile, home)}`;
  return writeChangedConfiguration(profile, original, updated);
}

export function updateProfiles(
  binDirectory: string,
  profiles: readonly string[],
  options: { home?: string } = {},
): [string[], [string, string][]] {
  const written: string[] = [];
  const failures: [string, string][] = [];
  for (const profile of profiles) {
    try {
      if (updateProfile(profile, binDirectory, options)) written.push(profile);
    } catch (error) {
      failures.push([profile, error instanceof Error ? error.message : String(error)]);
    }
  }
  return [written, failures];
}

export function removePathBlock(profile: string): boolean {
  let original: string;
  try {
    original = fs.readFileSync(profile, "utf8");
  } catch {
    return false;
  }
  let updated = original;
  for (;;) {
    const start = updated.indexOf(BLOCK_START);
    if (start === -1) break;
    const endMarker = updated.indexOf(BLOCK_END, start);
    if (endMarker === -1) break;
    let end = endMarker + BLOCK_END.length;
    if (updated[end] === "\n") end += 1;
    updated = updated.slice(0, start) + updated.slice(end);
  }
  return writeChangedConfiguration(profile, original, updated);
}

export function inspect(
  binDirectory: string,
  options: { home?: string; environment?: NodeJS.ProcessEnv; platformName?: string } = {},
): PathState {
  const home = options.home ?? os.homedir();
  const environment = options.environment ?? process.env;
  const onPath = isOnPath(binDirectory, { environment });
  const profiles = shellProfiles(options);
  let current = true;
  for (const profile of profiles) {
    try {
      const text = fs.readFileSync(profile, "utf8");
      if (!profileMentionsDirectory(text, binDirectory, home)) current = false;
    } catch {
      current = false;
    }
  }
  return {
    binDirectory,
    launcher: launcherPath(binDirectory, options.platformName),
    onPath,
    shadowedBy: shadowingExecutable(binDirectory, options),
    profiles,
    profilesCurrent: current,
  };
}

export function activationHint(
  profiles: readonly string[],
  options: { environment?: NodeJS.ProcessEnv } = {},
): string {
  const environment = options.environment ?? process.env;
  if (profiles.some((profile) => path.basename(profile) === "config.fish")) return "exec fish";
  const shell = environment.SHELL;
  return shell !== undefined && shell !== "" ? `exec ${shell} -l` : "exec $SHELL -l";
}
